Variance in `scipy.special.lambertw` output depending on scipy version and installation source

Originally discovered in https://github.com/pvlib/pvlib-python/pull/1996#issuecomment-2020819045

It seems that `scipy.special.lambertw` can return slightly different values from 1.12.0, depending on how it is installed (with conda or from PyPI with pip).  I observe this difference locally on Windows:

```python
>>> import scipy
>>> argW = 6.588940213030326e-08
>>> scipy.special.lambertw(argW)
(6.588939778889038e-08+0j)  # v1.12.0 (PyPI)
(6.588939778889037e-08+0j)  # v1.12.0 (conda)
```

As an aside, `mpmath` says that the one ending in `8` is the more accurate of the two:
```python
>>> import mpmath
>>> mpmath.mp.dps = 25
>>> mpmath.lambertw(argW)
mpf('0.0000000658893977888903808156892075')
```

The relevance is that the particular `argW` value above was (implicitly) used to generate negative values of `v_oc` using `_lambertw_v_from_i` for testing in `test_singlediode.py::test_singlediode_lambert_negative_voc`.  The change from `...8` to `...7` means that the returned voltage is no longer negative and thus causes the test to fail. 

Other information: `scipy.special.lambertw` was translated from cython to C++ for 1.12.0 in https://github.com/scipy/scipy/pull/19435.  

What should we do about it?  I think the first step is to ask the scipy folks if the difference is considered a bug in scipy that they will address.  I will ask the scipy maintainers about this.  If they say it is not a bug, then I suppose we should identify alternative values of `argW` that result in negative Voc for both versions of the scipy calculation and edit our test.

Credits to @cwhanse for the following comment:

> I can't reproduce this test error on my Windows system with scipy 1.12 and numpy 1.26.4. There were some changes to scipy's lambertw implementation for scipy 1.12 but I can't see that those changes would affect the values for that failed test. I don't know how to look to see if there's some Windows library that is different for scipy 1.10 vs. 1.12.

Locally it happens the same to me: no failures. However, on a remote Win11 machine, I have installed conda and what I find strange is that, compared to bare python, it is the only one that gives the following failure:

```
================================== FAILURES ===================================
____________________ test_singlediode_lambert_negative_voc ____________________

    def test_singlediode_lambert_negative_voc():
    
        # Those values result in a negative v_oc out of `_lambertw_v_from_i`
        x = np.array([0., 1.480501e-11, 0.178, 8000., 1.797559])
        outs = pvsystem.singlediode(*x, method='lambertw')
>       assert outs['v_oc'] == 0
E       assert 1.3234889800848443e-23 == 0

pvlib\tests\test_singlediode.py:175: AssertionError
```

Conda and Bare python installed environment packages versions listed and outputs after running the Conda and then Bare (scipy1.13.0), then restarting the remote and only doing Bare(sicpy1.13.0), then checking again with Bare(scipy1.12.0).

I think we can isolate this problem to only happen on conda, possibly due to some difference in their recipe/runner OS config in contrast to the scipy procedure. Anyways, that does not change the fact that these tests require some rework.

And by the way, there are a few warnings too. IMO we should strive to have clean tests.

--- a/pvlib/tests/test_singlediode.py
+++ b/pvlib/tests/test_singlediode.py
@@ -175,7 +175,7 @@ def test_singlediode_lambert_negative_voc():
     assert_approx_equal(outs['v_oc'], 0)
 
     # Testing for an array
-    x  = np.array([x, x]).T
+    x = np.array([x, x]).T
     outs = pvsystem.singlediode(*x, method='lambertw')
     assert_approx_equal(outs['v_oc'], [0, 0])
 

--- a/pvlib/tests/test_singlediode.py
+++ b/pvlib/tests/test_singlediode.py
@@ -177,15 +177,15 @@ def test_singlediode_lambert_negative_voc(mocker):
     # mock it. It depends on the platform and Python distro. See issue #2000.
     patcher = mocker.patch("pvlib.singlediode._lambertw_v_from_i")
     x = np.array([0.0, 1.480501e-11, 0.178, 8000.0, 1.797559])
-    patcher.return_value=-9.999e-13
+    patcher.return_value = -9.999e-13
     outs = pvsystem.singlediode(*x, method="lambertw")
     assert outs["v_oc"] == 0
 
     # Testing for an array
-    patcher.return_value=np.array([-9.999e-13, -1.001e-13])
+    patcher.return_value = np.array([-9.999e-13, -1.001e-13])
     x = np.array([x, x]).T
     outs = pvsystem.singlediode(*x, method="lambertw")
-    assert_array_equal(outs['v_oc'], [0, 0])
+    assert_array_equal(outs["v_oc"], [0, 0])
 
 
 @pytest.mark.parametrize('method', ['lambertw'])

--- a/pvlib/tests/test_singlediode.py
+++ b/pvlib/tests/test_singlediode.py
@@ -10,7 +10,7 @@ from pvlib.singlediode import (bishop88_mpp, estimate_voc, VOLTAGE_BUILTIN,
                                bishop88, bishop88_i_from_v, bishop88_v_from_i)
 from pvlib._deprecation import pvlibDeprecationWarning
 import pytest
-from numpy.testing import assert_allclose
+from numpy.testing import assert_array_equal
 from .conftest import DATA_DIR
 
 POA = 888
@@ -168,16 +168,24 @@ def test_singlediode_precision(method, precise_iv_curves):
     assert np.allclose(pc['i_xx'], outs['i_xx'], atol=1e-6, rtol=0)
 
 
-def test_singlediode_lambert_negative_voc():
-    # Those values result in a negative v_oc out of `_lambertw_v_from_i`
-    x = np.array([0., 1.480501e-11, 0.178, 8000., 1.797559])
-    outs = pvsystem.singlediode(*x, method='lambertw')
-    assert_allclose(outs['v_oc'], 0)
+def test_singlediode_lambert_negative_voc(mocker):
+    """Tests approximation to zero of v_oc when it is negative and small.
+    See singlediode.py:_lambertw > comment 'Set small elements <0 in v_oc to 0'
+    """
+    # Next values should result in a negative v_oc out of `_lambertw_v_from_i`
+    # however, we can't ensure that the output belongs to (-1e-12, 0), so we
+    # mock it. It depends on the platform and Python distro. See issue #2000.
+    patcher = mocker.patch("pvlib.singlediode._lambertw_v_from_i")
+    x = np.array([0.0, 1.480501e-11, 0.178, 8000.0, 1.797559])
+    patcher.return_value=-9.999e-13
+    outs = pvsystem.singlediode(*x, method="lambertw")
+    assert outs["v_oc"] == 0
 
     # Testing for an array
+    patcher.return_value=np.array([-9.999e-13, -1.001e-13])
     x = np.array([x, x]).T
-    outs = pvsystem.singlediode(*x, method='lambertw')
-    assert_allclose(outs['v_oc'].values, [0, 0])
+    outs = pvsystem.singlediode(*x, method="lambertw")
+    assert_array_equal(outs['v_oc'], [0, 0])
 
 
 @pytest.mark.parametrize('method', ['lambertw'])

--- a/pvlib/tests/test_singlediode.py
+++ b/pvlib/tests/test_singlediode.py
@@ -10,7 +10,7 @@ from pvlib.singlediode import (bishop88_mpp, estimate_voc, VOLTAGE_BUILTIN,
                                bishop88, bishop88_i_from_v, bishop88_v_from_i)
 from pvlib._deprecation import pvlibDeprecationWarning
 import pytest
-from numpy.testing import assert_approx_equal
+from numpy.testing import assert_allclose
 from .conftest import DATA_DIR
 
 POA = 888
@@ -172,12 +172,12 @@ def test_singlediode_lambert_negative_voc():
     # Those values result in a negative v_oc out of `_lambertw_v_from_i`
     x = np.array([0., 1.480501e-11, 0.178, 8000., 1.797559])
     outs = pvsystem.singlediode(*x, method='lambertw')
-    assert_approx_equal(outs['v_oc'], 0)
+    assert_allclose(outs['v_oc'], 0)
 
     # Testing for an array
     x = np.array([x, x]).T
     outs = pvsystem.singlediode(*x, method='lambertw')
-    assert_approx_equal(outs['v_oc'], [0, 0])
+    assert_allclose(outs['v_oc'].values, [0, 0])
 
 
 @pytest.mark.parametrize('method', ['lambertw'])

--- a/pvlib/tests/test_singlediode.py
+++ b/pvlib/tests/test_singlediode.py
@@ -10,6 +10,7 @@ from pvlib.singlediode import (bishop88_mpp, estimate_voc, VOLTAGE_BUILTIN,
                                bishop88, bishop88_i_from_v, bishop88_v_from_i)
 from pvlib._deprecation import pvlibDeprecationWarning
 import pytest
+from numpy.testing import assert_approx_equal
 from .conftest import DATA_DIR
 
 POA = 888
@@ -168,16 +169,15 @@ def test_singlediode_precision(method, precise_iv_curves):
 
 
 def test_singlediode_lambert_negative_voc():
-
     # Those values result in a negative v_oc out of `_lambertw_v_from_i`
     x = np.array([0., 1.480501e-11, 0.178, 8000., 1.797559])
     outs = pvsystem.singlediode(*x, method='lambertw')
-    assert outs['v_oc'] == 0
+    assert_approx_equal(outs['v_oc'], 0)
 
     # Testing for an array
     x  = np.array([x, x]).T
     outs = pvsystem.singlediode(*x, method='lambertw')
-    assert np.array_equal(outs['v_oc'], [0, 0])
+    assert_approx_equal(outs['v_oc'], [0, 0])
 
 
 @pytest.mark.parametrize('method', ['lambertw'])

